### DISCUSSION TITLE
ATTiny 816 PWM or PortMux issue?

### DISCUSSION THREAD

**Discussion by nipunawg** [SpenceKonde/megaTinyCore#782] [Q&A]
I've built an LED driver board with an integrated attiny816 (ext 16MHz crystal). The LED boost converter (CN5815) needs a PWM signal of 30KHz plus for dimming (I've tried analogwrite but it's buggy I'm guessing because the f is too low- it just turns off below about 150counts). I've connected PB5 to it's PWM input and set up the PWM as per the variable frequency example, but I'm not getting any PWM out of the pin.

My suspicion is that the PWM peripheral itself is running fine but I've set up portmux wrong. Anybody has any idea? Even if someone can provide me with some resources about portmux that would help a lot, because I can't find anything except what's on the datasheet but my implementation of that isn't working, obviously.

Below is the code I'm using to try and get an up/down brightness ramp to test (irrelevant code bits trimmed, full code attached)-
I'm initializing like so:
```
void setup(){
  pinMode(dimPin, OUTPUT);
  pinMode(enPin, OUTPUT);

  digitalWrite(enPin, LOW);       //turn on boost conv
  
  initPWM();                      //initialize pwm
  setFrequency(PWMFreq);          //set PWM frequency
  setDutyCycle(0);                //set output to pwm 0
}
```

and in the loop I'm calling:

```
void PWMRamp(){
  for(byte brightness=0; brightness<255; brightness++){
    setDutyCycle(brightness);
    delay(10);
  }
  for(byte brightness=255; brightness>1; brightness--){
    setDutyCycle(brightness);
    delay(10);
  }
}

```

And then I've got the PWM functions setup in a separate tab:
```

void initPWM() {                   
  takeOverTCA0();    
  PORTMUX.CTRLC    |= PORTMUX_TCA02_bm; // turn on PORTMUX, putting WO2 to PB5         
  TCA0.SINGLE.CTRLB = (TCA_SINGLE_CMP2EN_bm | TCA_SINGLE_WGMODE_SINGLESLOPE_gc);
  TCA0.SINGLE.PER   = Period; // Count all the way up to 0xFFFF
  //                             At 20MHz, this gives ~305Hz PWM
  TCA0.SINGLE.CMP0  = 0;
  TCA0.SINGLE.CTRLA = TCA_SINGLE_ENABLE_bm; // enable the timer with no prescaler
}

void setDutyCycle(byte duty) {
  TCA0.SINGLE.CMP0 = map(duty, 0, 255, 0, Period);
}

void setFrequency(unsigned long freqInHz) {
  unsigned long tempperiod = (F_CPU / freqInHz);
  byte presc = 0;
  while (tempperiod > 65536 && presc < 7) {
    presc++;
    tempperiod      = tempperiod >> (presc > 4 ? 2 : 1);
  }
  Period            = tempperiod;
  TCA0.SINGLE.CTRLA = (presc << 1) | TCA_SINGLE_ENABLE_bm;
  TCA0.SINGLE.PER   = Period;
}
```

full code:
[LED_FULLCTRL2 (2).zip](https://github.com/SpenceKonde/megaTinyCore/files/9436502/LED_FULLCTRL2.2.zip)

**Comment by grandaspanna** [comment]
There's a good chance you can use the standard analogWrite and keep things simple. I've got 38kHz working fine.

There was a recent discussion about this and this link has the succinct approach:
https://github.com/SpenceKonde/megaTinyCore/discussions/741#discussioncomment-3167576

**Comment by SpenceKonde** [comment]
Uhm, well, I see one bloody simple problem with your code. You're using WO2 as the output, but you're setting CMP0, changing the duty cycle for only WO0. That won't do anything to the duty cycle you're hoping to change. Change to writes to CMP0, which doesnt impact the pin you care about to CMP2 which does, and you should be getting much better results!
